accord-core/src/main/java/accord/impl/AbstractSafeCommandStore.java

```diff
@@ -85,6 +73,27 @@ private static <K, V> V getIfLoaded(K key, Function<K, V> get, Consumer<V> add,
         return value;
     }
 
+    protected abstract CommandType getCommandInternal(TxnId txnId);
+    protected abstract void addCommandInternal(CommandType command);
+    protected abstract CommandType getIfLoaded(TxnId txnId);
+
+    protected abstract TimestampsForKeyType getTimestampsForKeyInternal(RoutableKey key);
+    protected abstract void addTimestampsForKeyInternal(TimestampsForKeyType cfk);
+    protected abstract TimestampsForKeyType getTimestampsForKeyIfLoaded(RoutableKey key);
+
+    protected abstract CommandLoader<?> cfkLoader(RoutableKey key);
+    protected abstract CommandsForKeyType getDepsCommandsForKeyInternal(RoutableKey key);
```

Comment:
Do we really want to maintain three separate caches? I would be inclined to have a single cache entry that can contain any of the requested information (and if any is missing, to fetch only that) - I think it would be fine even if this didn't have separate eviction info, as we shouldn't often be touching the recovery set, and certainly the deps info and timestamp info are likely to be used proximally in time. So we would simply add anything missing at any point, while eviction might only be able to erase the whole set.

This would reduce the size of the changes I think, while also probably reducing the memory overhead and potentially increasing cache hit-rate.

WDYT?

Reply:
> Do we really want to maintain three separate caches?

It’s actually 4 separate caches :)

We have TimestampsForKey, CommandsForKey for both deps and recovery, and CommandsForKey.Updates, which allows us to issue cfk updates without having either set in memory.

> and certainly the deps info and timestamp info are likely to be used proximally in time.

It depends on how quickly commands are applied. With the current structure, the happy path only needs cfk data in memory for the Preaccept/Accept messages. After that, cfk data can be evicted and commands can complete through apply without having to reload anything, with the cfk updates being blindly buffered in memory until they’re flushed.

> So we would simply add anything missing at any point, while eviction might only be able to erase the whole set.

I’m having trouble parsing this sentence. Are you talking about loading missing info into the proposed multi-part cacheable state, or removing data from the deps cfk set? I don’t understand how eviction fits into either.

> This would reduce the size of the changes I think

It would reduce the changes on the accord side, but increase the changes and complexity on the C* side, and this was one of the main motivations for structuring things this way.

The AccordStateCache (in C*) has been kind of tricky to get right with single part state, and I’d like to avoid adding the notion of multi-part state to it if we can.

> probably reducing the memory overhead and potentially increasing cache hit-rate.

Fusing timestamps and deps commands would increase the cache rate, but also increase the memory overhead (though reduce the overall bookkeeping overhead)

I also think that some of the lower hanging fruit for more effective cache utilization will be making eviction smarter, ie: prioritizing eviction of recovery deps and buffered cfk updates, and keeping things split up this way makes that a bit easier.